[Spring MVC (인증)] 김진학 미션 제출합니다.

iampingu99 opened the pull request:
## 🏃‍♂️ 과정

이전 단계의 리뷰를 추가로 반영하고 새 단계의 요구사항을 추가했습니다. 역할 분리와 테스트에 대한 관점이 조금씩 만들어지는 것 같습니다. 이번 리뷰도 잘 부탁드립니다.

> `만료` 는 비즈니스 상 충분히 발생할 수 있는 요소라 생각합니다.
> 

> 이런 맥락에서 자신이 `이 내용은 여기서 테스트를 해야겠다.` `이 부분은 또 테스트를 할 필요 없겠는데.` 를 만들어가면 좋을거 같아요.
> 
- 테스트명 수정
- 토큰 서명 단위 테스트 삭제: 라이브러리의 일반적인(응답 비교) 검증은 제외
- 토큰 만료 단위 테스트 수정: 비즈니스 상 발생할 가능성이 높은 테스트 환경 조성을 위해 토큰 생성 객체에서 토큰 생성 날짜를 외부 객체에서 주입받아 역할을 분리
- 응답 바디 텍스트 검증 유지: 직관적인 메세지를 확인하기 위해 예상되는 예외가 올바른 메세지를 가지는지 확인

## **📃 요구사항**

### **로그인 리팩터링**

**로직**

- [x]  HandlerMethodArgumentResolver 를 사용하여 컨트롤러 진입 전 인증 정보 검증

### **예약 생성 기능 변경**

**로직**

- [x]  name이 없는 경우 Cookie에서 인증 정보 추출 후 예약하는 기능 추가

**예외**

- [x]  로그인을 하지 않은 경우
- [x]  예약 정보가 유효하지 않은 경우

### **관리자 기능**

**로직**

- [x]  /admin 으로 시작하는 모든 경로는 admin 권한이 있는 사람만 접근할 수 있도록 제한
- 컨트롤러 이전 단계에서 관리자 권한을 검증하는 HandlerIntercepter 과 로그인 객체를 생성하는 HandlerMethodArgumentResolver 동작 중, 쿠키에서 인증 토큰을 추출하고 객체를 생성하는 동작이 중복되었습니다. 이에 인증 토큰을 추출하는 역할은 AuthorizationExtractor 와 CookieAuthorizationExtractor 로 분리하게 되었고, 인증 객체를 생성하는 역할은 AuthService 에 두게 되었습니다.

## 🔎 궁금한점

로그인 상태가 필요한 모든 기능들에 대해서 로그인 요청 로직이 중복되었습니다. 이에 따라 아래와 같은 방법을 생각해보았습니다.

- 로그인과 같이 공통적으로 사용되는 로직을 객체로 분리하여 테스트에서 주입받는 방법: createToken 메서드의 경우 다른 곳에서도 사용될 가능성이 높다고 생각합니다.
- 이전에 작성한 테스트 객체를 주입받아 성공 메서드를 재사용하는 방법: 로그인 요청이 성공해야 하므로 로그인 성공 테스트 메서드를 사용할 수 있다고 생각합니다.

연쇄적으로 일어나야 하는 테스트에 대해서 이전 단계가 성공했다고 가정하고 단위 테스트로 쪼개거나, 통합 테스트에서 중복을 해결해야할 것으로 생각되는데 생각의 방향이 맞는지 궁금합니다.

iampingu99 committed:
refactor(auth): Cookie에 담긴 인증 정보를 이용해서 멤버 객체를 만드는 로직을 분리

iampingu99 committed:
refactor(auth): LoginMember 객체 생성 로직 팩토리 메서드로 추출

iampingu99 committed:
feat(reservation): 예약 생성 시 ReservationReqeust의 name이 없는 경우 Cookie에 담긴 정보를 활용

iampingu99 committed:
refactor(auth): 로그인 성공 테스트명 수정

iampingu99 committed:
refactor(auth): 토큰 테스트 필드 주입으로 수정

iampingu99 committed:
refactor(auth): 토큰 발행날짜 처리를 별도의 객체로 분리

만료된 토큰을 테스트하기 위해 토큰 생성 객체에서 날짜 처리를 분리해 발행날짜를 주입받도록 함

BREAKING CHANGE: 토큰 서명 테스트는 제거됨

iampingu99 committed:
test(reservation): 예약 기능 테스트 추가

iampingu99 committed:
test(reservation): 예약 정보 예외 테스트 추가

iampingu99 committed:
test(auth): 인증 정보 성공 테스트 추가

iampingu99 committed:
feat(auth): /admin 으로 시작하는 모든 경로는 admin 권한이 있는 사람만 접근할 수 있도록 제한

iampingu99 committed:
refactor(auth): 인증 토큰 추출 로직 외부 객체로 분리

로그인 확인 또는 어드민 권한 검사에 중복되는 인증 토큰 추출 로직을 외부 객체로 분리

iampingu99 committed:
refactor(auth): 인증 객체 생성 로직 분리

로그인 확인 또는 어드민 권한 검사에 중복되는 인증 객체 생성 로직을 AuthService 로 분리

iampingu99 committed:
refactor(auth): 인증 객체 정보 반환 타입 변경

키 값이 잘못 사용될 수 있는 Map 객체 대신 Claim 객체 사용

iampingu99 committed:
refactor(auth): 어드민 검사 구문 추출

Jackson은 isXXX 메서드를 JSON 필드로 간주하여 직렬화/역직렬화 작업에 포함하므로 @JsonIgnore 로 무시하도록
함

youngsu5582 commented on src/main/java/roomescape/auth/AdminInterceptor.java:
```diff
@@ -0,0 +1,32 @@
+import org.springframework.web.servlet.HandlerInterceptor;
+
+@Component
+public class AdminInterceptor implements HandlerInterceptor {
```
Interceptor 를 사용했네요.
Interceptor 와 Filter 의 차이점이 뭔가요?

iampingu99 replied:
잘못 이해하고 있거나, 이상한 부분은 말씀해주세요 👀

interceptor 는 dispatcherServlet 과 Controller 사이에서 요청이 controller에 도달하기 전 또는 후에 요청을 가로채게 됩니다. 요청이 들어왔을 때 dispatcherServlet 가 handlerMapping 을 통해서 적절한 controller 을 찾아서 매핑하게 되는데 이 직전에 로직을 수행할 수 있습니다. Filter 의 경우  dispatcherServlet 에 도달하기 전 또는 후에 요청을 가로채게 됩니다.

권한 검사의 경우 request 의 Cookie 인증 토큰을 통해서 권한을 검사하게 됩니다. 이 때, 권한 검사를 필요로 하는 요청이 많은 경우 해당 로직이 중복되게 됩니다. 이에, controllor 단의 로직 중복인 점을 생각해서 controller 직전에서 로직을 수행하는 것이 맞다고 생각하여 Intercepter 를 사용하게 되었습니다.

youngsu5582 replied:
그러면 Interceptor 와 Filter 는 둘다 스프링이 관리하나요?

iampingu99 replied:
아닙니다. Interceptor 는 스프링이 관리하며, Filter 는 웹서버의 일부입니다.

youngsu5582 commented on src/main/java/roomescape/auth/LoginMember.java:
```diff
@@ -0,0 +1,22 @@
+public record LoginMember(
+        String email,
+        String name,
+        String role
```
클린코드 관점에서 Role 이라는 ENUM 을 만들어도 될 거 같네요.

isAdmin 역시도 Role 에서 메소드로 가지고 있고 LoginMember 는 호출만 해도 되겠네용

iampingu99 replied:
필요성을 잘 못느꼈는데, Role 에 관한 유효성을 검증하는 메서드를 만들다 보니 코드가 엄청 지저분해지네요. ENUM 으로 관리한다면 훨씬 깔끔해질 것 같습니다. 수정해보겠습니다 👍

youngsu5582 commented on src/test/java/roomescape/reservation/CreateReservations.java:
```diff
@@ -0,0 +1,158 @@
+        reservationRequest.put("theme", "1");
+
+        //when
+        ReservationResponse reservationResponse = sendCreateReservationsRequest(reservationRequest, token).as(
```
직접 DTO 를 넘긴게 아니라 MAP 으로 넘겼는데 의도가 있을까요?
( 틀린게 아님. 의견 묻는거임 )

iampingu99 replied:
통합테스트가 실제 요청을 테스트하는 것 임으로 API 요청과 유사하게 key-value 쌍을 만들면 좋지 않을 까 라는 생각을 해보았습니다. 
이전 리뷰를 통해서 DTO의 필드 검증 측면에 대한 고민을 해본 결과, 필드 검증을 하려면 Map 객체보다는 DTO 객체를 사용해야 할 것 같습니다.

youngsu5582 replied:
좋은 접근이네요.
![image](https://github.com/user-attachments/assets/8b71715f-08f9-486a-9d47-8e8727612368)
https://toss.tech/article/test-strategy-server

해당 내용에 나와있는 부분을 적용해서 물어봤습니다. 
꼭 DTO 를 사용할 필요는 없지만, 현재 만들고 있는 사이드 프로젝트에서 위 방식으로 도입해봤는데
컴파일 타임 에러가 아니라, 런타임에 에러가 발생하는게 생각보다 매우 불편하더라구요. 🥲

이 역시도 나중에 팀원과 얘기를 하며 정해나갈 일종의 컨벤션일거 같아용

youngsu5582 commented on src/test/java/roomescape/reservation/CreateReservations.java:
```diff
@@ -0,0 +1,158 @@
+    @Test
+    void 비로그인_상태에서_예약할_경우_예약에_실패한다() {
+        //given
+        Map<String, String> reservationRequest = new HashMap<>();
```
request 를 만드는 부분도 private 메소드로 더 간편하게 해도 되겠네요.

iampingu99 replied:
Map 객체를 만드는 부분을 추출해서 사용해보란 말씀이신거죠? DTO와 Map 객체에 대한 추가적인 고민 후에 수정해보겠습니다 👍

iampingu99 replied:
> 이름/날짜/아이디 등등 형식이 맞지 않는 경우, 속성들의 키 와 값만 바뀌고 요청하는 로직은 모두 중복이 되게되었습니다.

> ReservationRequest 객체 수준의 테스트로 작성하려 했지만 내부 코드 오류보다는 "사용자의 잘못된 방식의 요청" 으로 인한 오류가 많다고 생각하여 통합 테스트로 작성하여 시간이 오래 걸리게 되었습니다.

위에 테스트 관해서 질문남긴 부분이 이 부분입니다.

1. 말씀 해주신 대로 private 메서드로 중복 로직을 제거하여 코드가 깔끔해졌습니다. 추가로 통합 테스트에서 `@ParameterizedTest` `@ValueSource` 어노테이션을 사용해서 하나의 테스트 메서드로 다양한 매개변수를 시도해보는 것은 어떻게 생각하시나요?
2. "사용자의 잘못된 입력" 검증으로 통합 테스트가 정확할 것 같아 이렇게 작성하게 되었습니다. 하지만 한편으로 reservationRequest dto 레이어에서 검사되어도 충분할 것 같다는 생각이 드는데 이런 상황에서는 기준을 어떻게 세우면 좋을까요?

youngsu5582 replied:
> "사용자의 잘못된 입력" 검증으로 통합 테스트가 정확할 것 같아 이렇게 작성하게 되었습니다. 하지만 한편으로 reservationRequest dto 레이어에서 검사되어도 충분할 것 같다는 생각이 드는데 이런 상황에서는 기준을 어떻게 세우면 좋을까요?

사실 해당 부분도 DTO 레이어 단에서 검증하고 끝낼거 같긴 합니다 저는.
테스트를 한다면 다다익선이긴 하나, 중복의 느낌이 들 거 같기 때문입니다.

근데, 그렇다고 해서 통합 테스트를 꼭 안해도 된다는건 아니에요.
어떤 JSON 구조 및 값으로 들어왔을때 문제없이 잘 동작하는가? 에 대해서도 검증할 필요가 있을수 있기 때문입니다.

은탄환은 없습니다! 🥲

youngsu5582 commented on src/main/java/roomescape/auth/AdminInterceptor.java:
```diff
@@ -0,0 +1,32 @@
+            throws Exception {
+
+        String token = authorizationExtractor.extract(request);
+        LoginMember loginMember = authService.createAuthentication(token);
```
`LoginMember.from()` 부분에서 Claims 에 해당하는 값이 없으면 예외가 터질거 같은데 이 경우 어떻게 되나요?

iampingu99 replied:
Jwt 의 payload 부분이 변조되는 경우예 예외가 발생할 것으로 예상했는데, 테스트 해본 결과 변조된 속성이 포함된 Claims 객체가 반환되는 것을 확인했습니다. 이에 두 가지 방법을 생각해보았습니다.

1. LoginMember 에 유효성 검사를 포함하여 생성자 함수 호출 시에 예외를 발생시키는 방법
  - 팩토리 메서드에서는 키 값이 존재하지 않는 경우 null을 넘겨준다
2. LoginMember 의 팩토리 메서드인 from 에서 필요한 키 값이 존재하지 않는 경우 예외를 발생시키는 방법

팩토리 메서드 보다는 생성자 함수에서 검사하는 것이 좀 더 사용성이 높아질 것으로 생각하는데, 혹시 다른 좋은 방법이 있을까요?

youngsu5582 replied:
제가 생각하기에는 해당 부분에서 굳이 팩토리 메소드가 필요할 까? 입니다.
특히, `Claims` 라는게 LoginMember 라는 객체가 가질 필요가 있는가? 싶기도 하구요.

Interceptor 가 쿠키에서 추출한 Claims 에서 필요한 요소들을 다 꺼내서 생성자로 만들면 될 거 같은데 어떤가요?

> 추가로, 팩토리 메소드에 검증을 넣는걸 되게 안좋게 생각합니다.
프로젝트에서도 어떤 곳은 검증을 한 팩토리 메소드를 사용하고, 어떤 곳은 생성자를 사용하니 굉장히 더럽더라구요.

iampingu99 replied:
LoginMember 가 인증된 사용자를 의미하는 객체이기 때문에 Claims 객체에 의존해도 충분하다는 생각이 들었습니다. 하지만 JWT 이외에 인증 방식을 사용하게 되면 수정이 필요하며, 이러한 상황은 가능성이 높습니다. 결과적으로 기본 생성자로도 충분히 객체 생성을 캡슐화할 수 있을 것 같습니다. 
추가로 아직은 다른 인증방식을 적용하지 않아서 분리의 필요성은 적지만, 인증객체를 생성하는 역할을 하는 인터페이스와 이를 구현하는 객체를 통해서 정리해볼 수도 있을 것 같습니다. 

저는 interceptor 보다는 service 레이어에서 필요한 요소들로 생성자를 호출하는 방식이 더 좋을 것 같습니다. 여러 인증방식을 추가하더라도 interceptor 의 관심사가 객체의 역할 검사인 만큼 수정을 하지 않는 것이 좋아보이는데 어떠실까요?

저 또한 팩토리 메소드 검증에 대한 의견에 동의합니다. 검증로직을 생성자 함수에 두어서 어떤 팩토리 메서드를 사용하더라도 마지막으로 생성자 함수를 호출하도록 하는 방식이 좋다고 생각합니다.

youngsu5582 commented on src/main/java/roomescape/auth/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,32 @@
+import org.springframework.web.method.support.ModelAndViewContainer;
+
+@Component
+public class LoginMemberArgumentResolver implements HandlerMethodArgumentResolver {
```
HandlerMethodArgumentResolver 가 정확하게 하는게 뭔가요?

iampingu99 replied:
잘못 이해하고 있거나, 이상한 부분은 말씀해주세요 👀

요청시에 사용된 매개변수들을 변환하기 위해서 사용됩니다. JSON 객체를 사용하기 위해서 역직렬화 과정이 필요하게 되는데 이를 수동으로 하는 것이 아닌 @RequestParam, @RequestBody 등의 어노테이션을 사용했을 때,  HandlerMethodArgumentResolver 활용하여 자동으로 매개변수에 매핑되게 됩니다.

하지만 요청 쿠키에서 토큰을 추출하고, 검증하여 검증 객체로 변환하는 로직은 자동으로 처리해주지 않았기 때문에, HandlerMethodArgumentResolver 를 직접 구현하게 되었습니다.

youngsu5582 commented on src/main/java/roomescape/reservation/ReservationController.java:
```diff
@@ -26,12 +26,16 @@ public List<ReservationResponse> list() {
-        if (reservationRequest.getName() == null
-                || reservationRequest.getDate() == null
+    public ResponseEntity create(@RequestBody ReservationRequest reservationRequest, LoginMember loginMember) {
+        if (reservationRequest.getDate() == null
```
물론, 기존 코드에서 검증을 여기서 하지만
DTO 가 스스로가 메소드로 자신의 값들이 타당한지`validXXX` 와 같은 메소드를 가지는것도 좋을거 같은데 어떤가요??

Controller 가 DTO 의 특정 요소가 null 인걸 알 필요가 있을까? 라고 생각해서 입니다.

iampingu99 replied:
ReservationRequest 객체 내에서 유효성 검사를 진행하는 것이 더 올바른 것 같습니다. 수정해보겠습니다. 👍 

추가로 객체의 모든 속성에 대해 검증을 하는 과정이 많을 것 같아, 테스트 코드를 한 번 작성해보게 되었습니다. 
1. 이름/날짜/아이디 등등 형식이 맞지 않는 경우, 속성들의 키 와 값만 바뀌고 요청하는 로직은 모두 중복이 되게되었습니다.
2. ReservationRequest 객체 수준의 테스트로 작성하려 했지만 내부 코드 오류보다는 "사용자의 잘못된 방식의 요청" 으로 인한 오류가 많다고 생각하여 통합 테스트로 작성하여 시간이 오래 걸리게 되었습니다.

테스트의 필요성을 느껴 작성했는데 자원을 필요 이상으로 사용하고 있는 것인지 제 방식이 마음에 들지 않습니다 ㅜ

youngsu5582 commented on src/main/java/roomescape/reservation/ReservationController.java:
```diff
@@ -26,12 +26,16 @@ public List<ReservationResponse> list() {
+            return ResponseEntity.badRequest().body("Invalid reservation request");
+        }
+
+        if (reservationRequest.getName() == null) {
```
이 부분도 위와 비슷하게 getName 이 null 인지 + 새로운 객체 생성을 DTO 내부로 은닉해도 될 거 같아요.

iampingu99 replied:
수정해보겠습니다 👍

youngsu5582 reviewed (changes requested):
우선 MVC 중점으로 리뷰 시작하겠습니다~

제가 남긴 의견들은 당연히 반박해주셔도 좋아요~
리뷰 반영하고 다시 멘션 주세요 🙂

youngsu5582 commented on src/test/java/roomescape/auth/LoginCheckTest.java:
```diff
@@ -6,12 +6,31 @@
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
 public class LoginCheckTest {
 
+    @Test
```
테스트를 했다는 부분이 여기 같은데 맞나요?

저 역시도 완벽한게 아니고, 정답은 없다고 생각합니다.

개인적인 의견으로 테스트는 필요한 부분까지만 하는게 최고라고 생각합니다.
코드는 고정된게 아니라 변화하는데
너무 엄격하게 모든걸 통제하면 변경을 방해하는 제약이 되더라구요.
( 저번에 말한것 처럼 프론트엔드 단의 요청 변경 )

테스트가 `해당 로직에서` 또는 `문서에서` 꼭 필요한 존재인지 생각하고 하면 될 거 같아요. 👍

지금 이 테스트 코드도 로젝에서나 문서에서 나쁘진 않은거 같아요.
자신만의 기준을 세워 나가 보시죠!

iampingu99 replied:
> 너무 엄격하게 모든걸 통제하면 변경을 방해하는 제약이 되더라구요.

테스트 기준을 마련하는데 큰 도움이 되는 말씀이신것 같습니다. 👍

youngsu5582 reviewed (changes requested):
달아주신 의견에 대해 제 생각 남겼습니다!
이제 MVC 에서 학습한 내용은 다 했다고 생각이 드는데
의견 보시고 더 궁금한 거 없으면 디스코드 다시 멘션주면 머지하겠습니다~ 👍

iampingu99 committed:
refactor(auth): Role 열거형 객체를 사용하여 역할 분리

iampingu99 committed:
refactor(auth): 의존성이 높은 팩토리 메서드 제외

의존성이 높은 팩토리 메서드는 재사용성 측면에서 이점을 볼 수 없다고 판단

BREAKING CHANGE: LoginMember.from(Claims) 메서드는 삭제됨

iampingu99 committed:
refactor(auth): 인증 객체 유효성 검사 추가

iampingu99 committed:
feat(auth): 토큰의 payload 형식이 일치하지 않는 예외 추가

iampingu99 committed:
refactor(reservations): 요청 객체 유효성 검사 추가

iampingu99 committed:
refactor(reservations): 예약 테스트 요청 공통 로직 추출

iampingu99 committed:
refactor(reservations): 예약 테스트 요청 공통 로직 추출

youngsu5582 commented on src/main/java/roomescape/auth/AuthService.java:
```diff
@@ -22,6 +22,9 @@ public String createToken(String email, String password) {
     public LoginMember createAuthentication(String token) {
         Claims claims = jwtTokenProvider.getClaims(token);
-        return LoginMember.from(claims);
+        return new LoginMember(
```
위에선 `jwtProvider` 를 쓰고, 아래에선 `JwtProvider` 를 쓰는 이유가 있나요?

iampingu99 replied:
> LoginMember.from() 부분에서 Claims 에 해당하는 값이 없으면 예외가 터질거 같은데 이 경우 어떻게 되나요?

위 리뷰로 인해서 잘못된 payload 로 인한 `NullPointException `예외를 확인해보았습니다. 또한 LoginMember 의 유효성 검사도 추가하여 `IllegalArgumentException` 를 발생시키도록 했습니다. 
이에 추출 역할을 어떻게 분리하면 좋을 지 고민하다, JwtProvider 에서 자주 사용되는 claim 추출 메서드를 정적으로 선언해서 구현하게 되었습니다.

youngsu5582 reviewed (approved):
진학님 이제 2~3단계 요구사항 및 질문점은 다 끝난거 같아서 Approve 하겠습니다.
하신 테스트들 사실 그냥 Java 의 문제라서...

질문 남긴거 하나만 다음에 반영해주세요!
